[BE]: 부서 CRUD, 사용자 CRUD

soyun-i opened the pull request:
<!-- 제목: [BE/FE/All] (기능 등) -->
<!-- resolves #에는 발행한 이슈 번호를 기입해주세요 -->

- resolves #2 

---

### 🚀 어떤 기능을 구현했나요 ?
- 사용자 CRUD
- 부서 CRUD
- 부서 목록 내려주기

### 🔥 어떤 문제를 마주했나요 ?
- enum 타입 Json 변환 과정에 문제가 있었습니다!

### ✨ 어떻게 해결했나요 ?
-  `@JsonCreator`활용하여 enum의 문자열 값을 enum 상수로 변환

### 📝 어떤 부분에 집중해서 리뷰해야 할까요?
- 

### 📚 참고 자료 및 회고 
<!--참고 자료(ref) 링크 및 스크린샷, 구현 과정에 대한 회고-->
- https://velog.io/@tin9oo/JsonCreate

soyun-i committed:
✨feature: 부서 CRUD, 사용자 CRUD 작성

soyun-i committed:
✨feature: 부서 CRUD, 사용자 CRUD 작성

soyun-i committed:
fix: 부서 삭제 테스트코드 수정

soyun-i committed:
fix: 부서 삭제 테스트코드 수정

JangBJ commented on src/main/java/com/kbe5/rento/domain/department/controller/DepartmentControllerImpl.java:
```diff
@@ -0,0 +1,59 @@
+
+@RestController
+@RequestMapping("/api/departments")
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
```
access를 붙이신 이유가 있으실까요?

soyun-i replied:
잘못 들어간 거 같습니다 수정하겠습니다!

JangBJ commented on src/main/java/com/kbe5/rento/domain/department/controller/DepartmentControllerImpl.java:
```diff
@@ -0,0 +1,59 @@
+    //부서 수정
+    @Override
+    @PutMapping("/{departmentId}")
+    public ResponseEntity<DepartmentInfoResponse> updateDepartment(@PathVariable Long departmentId, @Validated @RequestBody DepartmentUpdateRequest departmentUpdateRequest) {
```
코드 컨벤션 지켜주세요

JangBJ commented on src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
```
access를 붙이신 이유가 있으실까요?

JangBJ commented on src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java:
```diff
@@ -0,0 +1,100 @@
+@Service
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+public class DepartmentService {
+    private final DepartmentRepository departmentRepository;
```
개행 부탁드립니다

JangBJ commented on src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Transactional(readOnly = true)
+    public List<DepartmentInfoResponse> getDepartments(String companyCode) {
+        Company company = companyRepository.findByCompanyCode(companyCode).orElseThrow(() -> new DomainException(ErrorType.NO_SEARCH_RESULTS));
```
120자 넘어갈듯합니다? 코드 컨벤션 지켜주세요!

JangBJ commented on src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java:
```diff
@@ -0,0 +1,100 @@
+        List<Member> members = memberRepository.findAllByDepartmentId(departmentId);
+
+        if(!members.isEmpty()) {
+            throw new IllegalArgumentException("해당 부서에 소속된 직원이 있어 삭제할 수 없습니다.");
```
DomainExcepion을 사용해 주세요!

soyun-i replied:
현재 pull을 안해서 `IllegalArgumenrException`을 사용했습니다! pull 받아와서 다시 올리겠습니다!

JangBJ commented on src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java:
```diff
@@ -0,0 +1,100 @@
+
+    private void validateDuplicateDepartmentName(String departmentName, Long companyId) {
+        if(departmentRepository.existsByDepartmentNameAndCompanyId(departmentName, companyId)) {
+            throw new IllegalArgumentException("이미 존재하는 부서 이름입니다.");
```
DomainExcepion을 사용해 주세요!

JangBJ commented on src/main/java/com/kbe5/rento/domain/member/controller/MemberControllerImpl.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/members")
```
피라미드 형식 지켜주세요

SeungHun333 replied:
피라미드 규칙 적용해 주시면 될 것 같습니다!!

tmdgus717 replied:
피라미드형식입니다 🕶️

JangBJ commented on src/main/java/com/kbe5/rento/domain/member/controller/MemberControllerImpl.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Override
+    @PutMapping("/{memberId}")
+    public void update(@PathVariable Long memberId, @RequestBody @Validated MemberUpdateRequest request) {
```
코드 컨벤션 지켜주세요

JangBJ commented on src/main/java/com/kbe5/rento/domain/member/service/MemberService.java:
```diff
@@ -0,0 +1,96 @@
+        );
+
+        if(memberRepository.existsByPhoneNumberAndCompanyId(phoneNumber, company.getId())) {
+            throw new IllegalArgumentException("이미 존재하는 전화번호입니다.");
```
DomainException 사용해주세요!

SeungHun333 commented on src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java:
```diff
@@ -0,0 +1,100 @@
+
+@Service
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+public class DepartmentService {
```
이 클래스에서 첫 번째 줄 개행 안하신 것 같습니다!!

그리고 interface에도 안하신 부분 있어서 한번 확인 해주시면 될 것 같습니다!

tmdgus717 commented on src/main/java/com/kbe5/rento/domain/department/controller/DepartmentControllerImpl.java:
```diff
@@ -0,0 +1,59 @@
+
+@RestController
+@RequestMapping("/api/departments")
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
```
AccessLevel.PROTECTED 로 하신 이유가 궁금합니다~

soyun-i replied:
잘못 적은 거 같습니다 수정하겠습니다!

SeungHun333 commented on src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java:
```diff
@@ -0,0 +1,100 @@
+
+        departmentRepository.save(department);
+
+        return "성공적으로" + department.getDepartmentName() + "저장되었습니다";
```
return을 따로 DTO로 안하시고 바로 String값을 내려주는 것 보다 responseDTO로 내려주는 방식이 나중에 유지보수 할 때 더 수월하지 않을까요? 예를들어 부서 등록 시 저장되었다는 메시지 말고도 부서 아이디나 부서 정보 등을 내려줘야 할 때 바로 추가할 수 있을 것 같아서 말씀 드렸습니다!!

그리고 "성공적으로" + department.getDepartmentName() + "저장되었습니다"; 처럼 하드 코딩 한는 방식 보다 따로 상수를 만들고 DTO 메시지로 따로 보내는 방법도 괜찮아 보입니다!

soyun-i replied:
저장하는거라 void 로 할까 하다가 그래도 저장되는 것을 확인하기위해 String 으로 내려준것이었습니다! 승훈님 말씀처럼 String 으로 내려줄거면 상수처리를 하는 것이 더 좋을 거 같습니다! 일단 String 으로 내려주지 않고 body값에 "등록완료" 로 return 하게 수정하겠습니다! 

프론트 작업 진행하면서 필요한 값이 있을 시에 DTO 생성하여 사용하겠습니다!

SeungHun333 commented on src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java:
```diff
@@ -0,0 +1,100 @@
+        List<Member> members = memberRepository.findAllByDepartmentId(departmentId);
+
+        if(!members.isEmpty()) {
+            throw new IllegalArgumentException("해당 부서에 소속된 직원이 있어 삭제할 수 없습니다.");
```
"해당 부서에 소속된 직원이 있어 삭제할 수 없습니다." 부분도 하드코딩 보다는 따로 에러메시지 상수를 만들어서 적용하는 방법도 좋아 보입니다!

soyun-i replied:
맞습니다 ㅠ 예외처리 pull 받아와서 다시 수정하겠습니다!

tmdgus717 reviewed (approved):
LGTM👍

tmdgus717 commented on src/main/java/com/kbe5/rento/domain/department/dto/request/DepartmentRegisterRequest.java:
```diff
@@ -0,0 +1,19 @@
+                        .company(company)
+                        .build();
+        }
+}
```
아마 빈칸이없는 것 같은데요 정확히는 모르겠네요

soyun-i committed:
fix: 코드 컨벤션 수정 및 join column명 수정

soyun-i committed:
fix: 코드 컨벤션 수정 및 join column명 수정

# Conflicts:
#	src/main/java/com/kbe5/rento/domain/manager/entity/Manager.java
#	src/main/java/com/kbe5/rento/domain/manager/service/ManagerService.java

soyun-i committed:
fix: bcrypt 의존성 추가

soyun-i committed:
fix: DomainException 으로 변경

soyun-i committed:
Merge remote-tracking branch 'origin/feature/#2_사용자_CRUD' into feature/#2_사용자_CRUD

# Conflicts:
#	src/main/java/com/kbe5/rento/domain/department/service/DepartmentService.java
#	src/main/java/com/kbe5/rento/domain/manager/entity/Manager.java
#	src/main/java/com/kbe5/rento/domain/manager/service/ManagerService.java
#	src/main/java/com/kbe5/rento/domain/member/entity/Position.java
#	src/main/java/com/kbe5/rento/domain/member/service/MemberService.java

soyun-i committed:
fix: password length 수정 취소

soyun-i committed:
fix: 테스트 코드 수정

soyun-i committed:
fix: 테스트 코드 수정

tmdgus717 reviewed (approved):
👍
